pySDC/playgrounds/Diagonal/dahlquist.py

```diff
@@ -0,0 +1,350 @@
+
+        # Update parameters
+        if updateNode:
+            cls.nodes, cls.weights, cls.Q = genCollocation(
+                M, nodeDistr, quadType)
+            cls.nodeDistr, cls.quadType = nodeDistr, quadType
+        if updateQDeltaI:
+            cls.QDeltaI, cls.dtauI = genQDelta(cls.nodes, implSweep, cls.Q)
+            cls.implSweep = implSweep
+        if updateQDeltaE:
+            cls.QDeltaE, cls.dtauE = genQDelta(cls.nodes, explSweep, cls.Q)
+            cls.explSweep = explSweep
+
+        # Eventually update nSweep, initSweep and forceProlongation
+        cls.initSweep = cls.initSweep if initSweep is None else initSweep
+        cls.nSweep = cls.nSweep if nSweep is None else nSweep
+        cls.forceProl = cls.forceProl if forceProl is None else forceProl
+
+    @classmethod
+    def getMaxOrder(cls):
+        # TODO : adapt to non-LEGENDRE node distributions
+        M = len(cls.nodes)
+        return 2*M if cls.quadType == 'GAUSS' else \
+            2*M-1 if cls.quadType.startswith('RADAU') else \
+            2*M-2  # LOBATTO
+
+    def __init__(self, u0, lambdaI, lambdaE):
+
+        model = np.asarray(np.asarray(lambdaI) + np.asarray(lambdaE) + 0.)
+        c = lambda : np.zeros_like(model)
+
+        self.lambdaI, self.lambdaE = c(), c()
+        np.copyto(self.lambdaI, lambdaI)
+        np.copyto(self.lambdaE, lambdaE)
+
+        self.u = c()  # Equivalent to solver state
+        np.copyto(self.u, u0)  # Initialize state
+
+        self.rhs, self.u0 = c(), c()
+        self.lamIU = deque([[c() for _ in range(self.M)] for _ in range(2)])
+        self.lamEU = deque([[c() for _ in range(self.M)] for _ in range(2)])
+        if not self.leftIsNode:
+            self.lamEU0, self.lamIU0 = c(), c()
+
+        # Instanciate list of solver (ony first time step)
+        self.lhs = [None] * self.M
+
+        self.dt = None
+        self.axpy = blas.get_blas_funcs('axpy', dtype=self.u.dtype)
+
+    @property
+    def M(self):
+        return len(self.nodes)
+
+    @property
+    def rightIsNode(self):
+        return np.allclose(self.nodes[-1], 1.0)
+
+    @property
+    def leftIsNode(self):
+        return np.allclose(self.nodes[0], 0.0)
+
+    @property
+    def doProlongation(self):
+        return not self.rightIsNode or self.forceProl
+
+    def _storeU0(self):
+        np.copyto(self.u0, self.u)
+
+    def _updateLHS(self, dt, init=False):
+        """Update LHS coefficients
+
+        Parameters
+        ----------
+        dt : float
+            Time-step for the updated LHS.
+        init : bool, optional
+            Wether or not initialize the LHS_solvers attribute for each
+            subproblem. The default is False.
+        """
+        # Attribute references
+        qI = self.QDeltaI
+        for i in range(self.M):
+            self.lhs[i] = 1 - dt*qI[i, i]*self.lambdaI
+
+    def _evalImplicit(self, lamIU):
+        np.copyto(lamIU, self.u)
+        lamIU *= self.lambdaI
```

Comment:
It avoids data duplication : if you do

```python
lamIU = self.u * self.lambdaI
```
then each time you call the _evalImplicit method, you build a new array (that is the product between `u` and `lambdaI`), and assign `lamIU` to it. 

In the proposed approach, `lamIU` is a pre-allocated array that is built only when the SDC solver is instantiated, and only its values are modified during computation. It's a bit like the _efficient_ approach you would consider when programming in C++ : never use copy constructor, instantiate all the array you need before computation, and only modify their value during computation.

Of course, for this example it does not make a big difference, since the considered arrays are not very large. But I would say it's just trying to keep some _good habits when using Python for Scientific Computing_ :smile:

Reply:
Interesting! Thanks!